Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,180 @@
+//
+//  Created by Robin Macharg on 04/05/2020.
+//  Copyright © 2020 Bugsnag. All rights reserved.
+//
+
+#import <mach-o/dyld.h>
+#import <dlfcn.h>
+#import <Foundation/Foundation.h>
+#import "BSG_KSDynamicLinker.h"
+#import "BSG_KSMachHeaders.h"
+
+// We expect <1000 items
+static const int BSG_INITIAL_MACH_BINARY_IMAGE_ARRAY_SIZE = 100;
+
+void bsg_initialize_binary_images_array(size_t initialSize) {
+    bsg_mach_binary_images.contents = (BSG_Mach_Binary_Image_Info *)malloc(initialSize * sizeof(BSG_Mach_Binary_Image_Info));
+    bsg_mach_binary_images.used = 0;
+    bsg_mach_binary_images.size = initialSize;
+}
+
+BSG_Mach_Binary_Images *bsg_get_mach_binary_images() {
+    return &bsg_mach_binary_images;
+}
+
+/**
+ * Store a Mach binary image-excapsulating struct in a dynamic array.
+ * The array doubles on filling-up.  Typical sizes is expected to be in < 1000 (i.e. 2-3 doublings, at app start-up)
+ * This should be called in a threadsafe way; we lock against a simultaneous add and remove.
+ */
+void bsg_add_mach_binary_image(BSG_Mach_Binary_Image_Info element) {
+    
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    
+    // Expand array if necessary
+    if (bsg_mach_binary_images.used == bsg_mach_binary_images.size) {
+        bsg_mach_binary_images.size *= 2;
+        bsg_mach_binary_images.contents = (BSG_Mach_Binary_Image_Info *)realloc(bsg_mach_binary_images.contents, bsg_mach_binary_images.size * sizeof(BSG_Mach_Binary_Image_Info));
+    }
+    
+    // Store the value, increment the number of used elements
+    bsg_mach_binary_images.contents[bsg_mach_binary_images.used++] = element;
+    
+    os_unfair_lock_unlock(&bsg_mach_binary_images_access_lock);
+}
+
+/**
+ * Binary images can only be loaded at most once.  We can use the (file)name as a key, without needing to compare the
+ * other fields.  Element order is not important; deletion is accomplished by copying the last item into the deleted
+ * position.
+ */
+void bsg_remove_mach_binary_image(const char *element_name) {
+    
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock);
+    
+    for (size_t i=0; i<bsg_mach_binary_images.used; i++) {
+        BSG_Mach_Binary_Image_Info item = bsg_mach_binary_images.contents[i];
+        
+        if (strcmp(element_name, item.name) == 0) {
+            // Note: removal of the last (ith) item involves a redundant copy from last->last.
+            if (bsg_mach_binary_images.used >= 2) {
+                bsg_mach_binary_images.contents[i] = bsg_mach_binary_images.contents[--bsg_mach_binary_images.used];
+            }
+            else {
+                bsg_mach_binary_images.used = 0;
+            }
+            break; // an image can only be loaded singularly; exit loop once found
+        }
+    }
+    
+    os_unfair_lock_unlock(&bsg_mach_binary_images_access_lock);
+}
+
+void bsg_initialise_mach_binary_headers() {
+    bsg_initialize_binary_images_array(BSG_INITIAL_MACH_BINARY_IMAGE_ARRAY_SIZE);
+}
```

Comment:
Nit: do we still need this extra function?

Reply:
This facilitated easier testing but, if you're happy for the default array size value to be delocalised slightly and live in `BSG_KSCrash.m`, then we can do away with the extra level.

Reply:
No strong view - your call.